### PR TITLE
More Server Variable/Options & Slight Rubocop changes

### PR DESCRIPTION
**Pull request by SnazzyPine25**

### CODE REVIEW

File: lib/discordrb/data.rb [hunk 1]
```diff
@@ -2312,6 +2325,10 @@ def initialize(data, bot, exists = true)
       @large = data['large']
       @member_count = data['member_count']
       @verification_level = [:none, :low, :medium, :high][data['verification_level']]
+      @explicit_content_filter = [:none, :exclude_roles, :all][data['explicit_content_filter']]
+      @default_message_notifications = [:all, :mentions][data['default_message_notifications']]
+      @widget_enabled = data['widget_enabled']
```
Comment (meew0) [inline lib/discordrb/data.rb]: These properties are actually called `embed_enabled` and `embed_channel_id` in the data JSON.

**TEST
YOUR
CODE
PLEASE**

File: lib/discordrb/data.rb [hunk 2]
```diff
@@ -2312,6 +2325,10 @@ def initialize(data, bot, exists = true)
...
+      @widget_channel = data['widget_channel_id'].nil? ? nil : bot.ensure_channel(data['channels'][data['widget_channel_id']], self)
```
Comment (meew0) [inline lib/discordrb/data.rb]: Also, could you move these assignments you made into `update_data`? That way they should get updated on `GUILD_UPDATE` as well.

File: lib/discordrb/data.rb [hunk 3]
```diff
@@ -2293,9 +2293,22 @@ class Server
...
+    # @return [Symbol] the default message notifications settings of the server (:all = 'All messages', :mentions = 'Only @mentions').
+    attr_reader :default_message_notifications
+
     # @return [Integer] the amount of time after which a voice user gets moved into the AFK channel, in seconds.
     attr_reader :afk_timeout
 
+    # @return [true, false] whether or not the widget is enabled on the server.
+    attr_reader :widget_enabled
+
+    # @return [Channel, nil] the channel the server widget will make a invite for.
```
Comment (meew0) [inline lib/discordrb/data.rb]: `an invite`

File: lib/discordrb/data.rb [hunk 4]
```diff
@@ -2293,9 +2293,22 @@ class Server
...
+    alias_method :content_filter_level, :explicit_content_filter
+
+    # @return [Symbol] the default message notifications settings of the server (:all = 'All messages', :mentions = 'Only @mentions').
+    attr_reader :default_message_notifications
+
     # @return [Integer] the amount of time after which a voice user gets moved into the AFK channel, in seconds.
     attr_reader :afk_timeout
 
+    # @return [true, false] whether or not the widget is enabled on the server.
+    attr_reader :widget_enabled
```
Comment (meew0) [inline lib/discordrb/data.rb]: Please add a `widget_enabled?` alias for this one, as it is a boolean property

File: lib/discordrb/data.rb (hunk 1 above)
Comment (SnazzyPine25) [inline lib/discordrb/data.rb]: **W R O N G**

File: lib/discordrb/data.rb [hunk 5]
```diff
@@ -2407,13 +2403,30 @@ def integrations
     # @!visibility private
     def cache_embed
       @embed ||= JSON.parse(API::Server.resolve(@bot.token, @id))['embed_enabled']
+      @embed_channel_id ||= JSON.parse(API::Server.resolve(@bot.token, @id))['embed_channel_id']
```
Comment (z64) [inline lib/discordrb/data.rb]: Why are you calling this `JSON.parse #..` expression twice? 🤔

File: lib/discordrb/data.rb [hunk 6]
```diff
@@ -2312,6 +2321,10 @@ def initialize(data, bot, exists = true)
       @large = data['large']
       @member_count = data['member_count']
       @verification_level = [:none, :low, :medium, :high][data['verification_level']]
+      @explicit_content_filter = [:none, :exclude_roles, :all][data['explicit_content_filter']]
+      @default_message_notifications = [:all, :mentions][data['default_message_notifications']]
+      @widget_enabled = data['widget_enabled']
+      @embed_channel = nil
```
Comment (meew0) [inline lib/discordrb/data.rb]: These assignments no longer need to be here as they're handled by `update_data`.

File: lib/discordrb/data.rb [hunk 7]
```diff
@@ -2259,6 +2259,8 @@ def inspect
   end
 
   # A server on Discord
+
+  # A server on Discord
```
Comment (meew0) [inline lib/discordrb/data.rb]: What's with the extra comment?

File: lib/discordrb/data.rb [hunk 8]
```diff
@@ -2389,14 +2402,32 @@ def integrations
...
       cache_embed if @embed.nil?
       @embed
     end
+    alias_method :widget_enabled, :embed?
+    alias_method :widget?, :embed?
+    alias_method :embed_enabled, :embed?
+
+    # @return [Channel, nil] the channel the server embed will make a invite for.
+    def embed_channel
+      cache_embed if @embed_channel.nil?
```
Comment (meew0) [inline lib/discordrb/data.rb]: This check should check whether `@embed` is nil, because the `@embed_channel` might also be nil if the embed is cached but the channel is unreachable.

File: spec/data/emoji/emoji_server.json [hunk 9]
```diff
@@ -1,5 +1,9 @@
 {
   "verification_level": 0,
+  "explicit_content_filter": 0,
+  "default_message_notifications": 0,
+  "widget_enabled": false,
+  "widget_channel_id": null,
```
Comment (meew0) [inline spec/data/emoji/emoji_server.json]: You're checking for `embed_` properties in the code, but here only the `widget_` ones are present - you might as well put both variants in, as Discord does.

File: lib/discordrb/data.rb [hunk 10]
```diff
@@ -2290,9 +2290,16 @@ class Server
     # @return [Integer] the absolute number of members on this server, offline or not.
     attr_reader :member_count
 
-    # @return [Symbol] the verification level of the server (:none = none, :low = 'Must have a verified email on their Discord account', :medium = 'Has to be registered with Discord for at least 5 minutes', :high = 'Has to be a member of this server for at least 10 minutes').
+    # @return [Symbol] the verification level of the server (:none = none, :low = 'Must have a verified email on their Discord account', :medium = 'Has to be registered with Discord for at least 5 minutes', :high = 'Has to be a member of this server for at least 10 minutes', :extra = 'Must have a verified phone ob their Discord account').
```
Comment (meew0) [inline lib/discordrb/data.rb]: I don't really like the name `extra`, how about `very_high`?

Comment (meew0) [inline lib/discordrb/data.rb]: Also, `on their Discord account`

File: lib/discordrb/data.rb [hunk 11]
```diff
@@ -2290,9 +2290,16 @@ class Server
...
     attr_reader :verification_level
 
+    # @return [Symbol] the explicit content filter level of the server (:none = 'Don't scan any messages', :exclude_roles = 'Scan messages for members without a role.', :all = 'Scan messages sent by all messages.').
```
Comment (meew0) [inline lib/discordrb/data.rb]: `Scan messages sent by all members.`; also `Don't scan any messages` needs a period at the end

File: lib/discordrb/data.rb [hunk 12]
```diff
@@ -2311,7 +2318,10 @@ def initialize(data, bot, exists = true)
 
       @large = data['large']
       @member_count = data['member_count']
-      @verification_level = [:none, :low, :medium, :high][data['verification_level']]
+      @verification_level = [:none, :low, :medium, :high, :extra][data['verification_level']]
```
Comment (meew0) [inline lib/discordrb/data.rb]: These four assignments are already handled by `update_data` and aren't necessary here

File: lib/discordrb/data.rb [hunk 13]
```diff
@@ -2719,15 +2747,26 @@ def update_data(new_data)
       @region_id = new_data[:region] || new_data['region'] || @region_id
       @icon_id = new_data[:icon] || new_data['icon'] || @icon_id
       @afk_timeout = new_data[:afk_timeout] || new_data['afk_timeout'].to_i || @afk_timeout
-
       @afk_channel_id = new_data[:afk_channel_id] || new_data['afk_channel_id'].to_i || @afk_channel.id
+      @embed_channel_id = new_data[:embed_channel_id] || new_data['embed_channel_id'].to_i || @embed_channel.id
+      @embed = new_data[:embed_enabled] || new_data['embed_enabled'] || @embed
+      @verification_level = [:none, :low, :medium, :high][new_data['verification_level']] || @verification_level
```
Comment (meew0) [inline lib/discordrb/data.rb]: This is missing the `extra` (or `very_high`) verification level

File: lib/discordrb/data.rb [hunk 14]
```diff
@@ -2594,20 +2618,23 @@ def bans
...
+      API::Server.ban_user(@bot.token, @id, user.resolve_id, message_days, reason)
     end
 
     # Unbans a previously banned user from this server.
     # @param user [User, #resolve_id] The user to unban.
-    def unban(user)
-      API::Server.unban_user(@bot.token, @id, user.resolve_id)
+    # @param reason [String] The reason the user is being unbanned.
+    def unban(user, reason = nil)
+      API::Server.unban_user(@bot.token, @id, user.resolve_id, reason = nil)
```
Comment (z64) [inline lib/discordrb/data.rb]: Isn't this throwing away any reason the user passed to the method...?

File: lib/discordrb/data.rb [hunk 15]
```diff
@@ -2594,20 +2618,23 @@ def bans
...
     end
 
     # Kicks a user from this server.
     # @param user [User, #resolve_id] The user to kick.
-    def kick(user)
-      API::Server.remove_member(@bot.token, @id, user.resolve_id)
+    # @param reason [String] The reason the user is being kicked.
+    def kick(user, reason = nil)
+      API::Server.remove_member(@bot.token, @id, user.resolve_id, reason = nil)
```
Comment (z64) [inline lib/discordrb/data.rb]: Same as above?

File: spec/data/emoji/emoji_server.json [hunk 16]
```diff
@@ -18,4 +24,4 @@
       "id": "11"
     }
   ]
-}
\ No newline at end of file
+}
```
Comment (Daniel-Worrall) [inline spec/data/emoji/emoji_server.json]: Why was this newline added here? The other data files for specs don't include this, and it doesn't do anything in regards to the PR

File: lib/discordrb/data.rb [hunk 17]
```diff
@@ -2311,7 +2318,6 @@ def initialize(data, bot, exists = true)
 
       @large = data['large']
       @member_count = data['member_count']
-      @verification_level = [:none, :low, :medium, :high][data['verification_level']]
```
Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: Why was this removed?

File: lib/discordrb/data.rb [hunk 18]
```diff
@@ -2594,20 +2618,23 @@ def bans
     # Bans a user from this server.
     # @param user [User, #resolve_id] The user to ban.
     # @param message_days [Integer] How many days worth of messages sent by the user should be deleted.
-    def ban(user, message_days = 0)
-      API::Server.ban_user(@bot.token, @id, user.resolve_id, message_days)
+    # @param reason [String] The reason the user is being banned.
+    def ban(user, message_days = 0, reason = nil)
```
Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: Perhaps rework this to use named parameters so that a reason can be specified without specifying the days

**Comment by Daniel-Worrall** [comment]
Can't help but notice this title was changed to include a typo. More, not Mose.

File: lib/discordrb/api/channel.rb [hunk 19]
```diff
@@ -16,27 +16,29 @@ def resolve(token, channel_id)
 
   # Update a channel's data
   # https://discordapp.com/developers/docs/resources/channel#modify-channel
-  def update(token, channel_id, name, topic, position, bitrate, user_limit)
+  def update(token, channel_id, name, topic, position, bitrate, user_limit, reason)
```
Comment (z64) [inline lib/discordrb/api/channel.rb]: All of these should probably have `nil` as a default for reason. But, more importantly, you specify a default for some, but not others - let's pick one and stick with it. Or at least try to be *more* consistent.

File: lib/discordrb/api/channel.rb [hunk 20]
```diff
@@ -111,27 +113,29 @@ def edit_message(token, channel_id, message_id, message, mentions = [], embed =
 
   # Delete a message
   # https://discordapp.com/developers/docs/resources/channel#delete-message
-  def delete_message(token, channel_id, message_id)
+  def delete_message(token, channel_id, message_id, reason)
```
Comment (z64) [inline lib/discordrb/api/channel.rb]: This action is not logged for bots. It would probably be best to omit this.

File: lib/discordrb/api/channel.rb [hunk 21]
```diff
@@ -111,27 +113,29 @@ def edit_message(token, channel_id, message_id, message, mentions = [], embed =
...
-      Authorization: token
+      Authorization: token,
+      'X-Audit-Log-Reason': reason
     )
   end
 
   # Delete messages in bulk
   # https://discordapp.com/developers/docs/resources/channel#bulk-delete-messages
-  def bulk_delete_messages(token, channel_id, messages = [])
+  def bulk_delete_messages(token, channel_id, messages = [], reason = nil)
```
Comment (z64) [inline lib/discordrb/api/channel.rb]: This action does not emit an audit log. Please test that all you added `reason` to actually emits logs.

https://github.com/hammerandchisel/discord-api-docs/issues/265#issuecomment-305307585

File: lib/discordrb/data.rb [hunk 22]
```diff
@@ -1454,12 +1474,14 @@ def position=(position)
     # permission sets, or change an existing one.
     # @overload define_overwrite(overwrite)
     #   @param thing [Overwrite] an Overwrite object to apply to this channel
+    #   @param reason [String] The reason the for the overwrite.
```
Comment (z64) [inline lib/discordrb/data.rb]: "The reason for defining the overwrite" or "The reason for creating the overwrite"

File: lib/discordrb/data.rb [hunk 23]
```diff
@@ -1454,12 +1474,14 @@ def position=(position)
...
     # @overload define_overwrite(thing, allow, deny)
     #   @param thing [User, Role] What to define an overwrite for.
     #   @param allow [#bits, Permissions, Integer] The permission sets that should receive an `allow` override (i. e. a
     #     green checkmark on Discord)
     #   @param deny [#bits, Permissions, Integer] The permission sets that should receive a `deny` override (i. e. a red
     #     cross on Discord)
+    #   @param reason [String] The reason the for the overwrite.
```
Comment (z64) [inline lib/discordrb/data.rb]: ditto

**Comment by Daniel-Worrall** [comment]
You've changed the Ruby version on both Travis and Rubocop, but you haven't updated the gemspec.
